userspace/libscap/scap_suppress.h

```diff
@@ -46,11 +53,18 @@ struct scap_suppress
 	// The number of events that were skipped due to the comm
 	// matching an entry in m_suppressed_comms.
 	uint64_t m_num_suppressed_evts;
+
+	// The cache around the m_suppressed_tids hash table. Why?
+	// In benchmarks as of June 2023, scap_check_suppressed() with
+	// suppressed TIDs increases speed from 17.51s to 7.24s. Why?
+	// The cache is a single cache line fetch with no hashing.
+	scap_tid_stid m_cpuid_tid_stid_cache[SCAP_CPUID_MAX];
```

Comment:
Why do we want to tie the number of cache entries to `*pcpuid`? `scap_next` runs in a single thread and the *pcpuid value you use to choose the cache entry is the cpu the event was generated on (for kmod at least, for other engines it may well be meaningless or at least unrelated to physical cpus), not the cpu the event is consumed on.

So, IMHO:
* don't treat `*pcpuid` as anything CPU related
* pick a way smaller SCAP_CPUID_MAX value
* use `*pcpuid % SCAP_CPUID_MAX` as the index
* rename all the `cpu` references to something less specific (`cache`?)

Also, how does the benchmark look with just a single cache entry?

Reply:
Actually, using the (low bits of) tid instead of cpuid might make more sense, we'd have some form of cache associativity then :)

Reply:
(then let's add some chaining, or maybe open addressing... :trollface:)

Reply:
Thanks you so much for all the comments and good questions! :-)

> Why do we want to tie the number of cache entries to `*pcpuid`?

This is an idea which came from Loris after I presented the same thing not using `*pcpuid`:

> when I originally wrote the sinsp thread table, I implemented a simple LRU cache that avoided any lookup in the table if the TID didn't change compared to the previous event. This, at that time, was very effective at reducing the number of TID lookups

I preferred this variation because it uses many less active cache lines for boxes with fewer CPUs, although it is a similar speed improvement at run-time.

The original implementation benchmarked did not use `*pcpuid` and instead cached by e.g. `tid & 1023`.

> *pcpuid value you use to choose the cache entry is the cpu the event was generated on

That's right. The idea is that the stream of events from a particular CPU often come from the same TID (depending on the amount of system context switching).

> use *pcpuid % SCAP_CPUID_MAX as the index

The `%` operator is relatively heavy-weight compared to e.g. `&`, e.g. in this example it's 30 times slower:

```
$ cat foo.c
#include <stdio.h>
#include <stdint.h>
void main(void) {
    uint64_t v = 12345678901234567890UL;
    uint64_t r = 0;
    for(uint64_t i = 1; i <= 1000000000; i ++) {
        r += v OPERATOR i;
    }
    printf("- r=%ld\n", r);
}
$ gcc -DOPERATOR="&" -O1 -o foo.exe foo.c && time ./foo.exe
...
real	0m0.299s
$ gcc -DOPERATOR="%" -O1 -o foo.exe foo.c && time ./foo.exe
...
real	0m8.987s
```

> pick a way smaller SCAP_CPUID_MAX value

If we reduce the value to e.g. 128 then boxes with > 128 CPUs will not benefit from this optimization.
Why? If TIDs are running on the "same" cache slot they will continually evict each others TIDs.
Currently the memory overhead is 1024 * 16 bytes = 16 KB.
Although this seems like a lot, optimization is for the inner most loop and helps every event.
In benchmarks the 16 KB helps claw back several CPU % points at run-time, e.g. saving > 2% total CPU at run-time.

> Also, how does the benchmark look with just a single cache entry?

Not sure I follow the logic here.
A "single cache entry" would get evicted nearly all the time and provide almost no benefit, or?

> Actually, using the (low bits of) tid instead of cpuid might make more sense, we'd have some form of cache associativity then :)

Yes, using `&` for the low bits was used in the original experimental implementation.
It works okay but you end up needing to use many TID cache slots, instead of CPU cache slots. Why?
Because there are generally many less CPUs than TIDs on a box.
So caching with e.g. 32 CPU cache slots gives the same performance as caching with 1024 TID cache slots.
And we don't know how many TIDs will be running on a system.
I have used the Google Microservices Demo for benchmarking. It's relatively "small" but uses ~ 800 TIDs.

> (then let's add some chaining, or maybe open addressing... :trollface:)

The main performance benefit comes from doing less cache line fetches.
However, adding chaining or open addressing will only increase the number of cache line fetches.
Currently the cache has a single chance to get its answer at a single cache line.
If the answer is not there then it reverts to the hash table which does its chaining, etc.

Reply:
> Why do we want to tie the number of cache entries to *pcpuid? scap_next runs in a single thread and the *pcpuid value you use to choose the cache entry is the cpu the event was generated on (for kmod at least, **for other engines it may well be meaningless or at least unrelated to physical cpus**)

That's a very good point about the "other engines"! AFAIK "other engines" means the eBPF drivers, or? Do they report the CPU the event was generated on too? If yes, I think we're okay. If not, then sorry for the confusion, and I can revert to the version of the algorithm which uses the TID to calculate the cache slot, instead of using the CPU ID. And if we don't know, who would know? Somebody suggested that @Andreagit97 might know? :-)

Reply:
> The `%` operator is relatively heavy-weight compared to e.g. `&`

I was silently assuming the right-hand side would be a constant power of 2, so the compiler would generate `&` anyway :)

> A "single cache entry" would get evicted nearly all the time and provide almost no benefit, or?

I'd just want to see how it would hold up. Of course it would get evicted more often, I'm just wondering how bad it would be (we have similar single-entry caches elsewhere)

> AFAIK "other engines" means the eBPF drivers, or? Do they report the CPU the event was generated on too? 

I'm not sure what modern_bpf does when the number of buffers doesn't match the number of CPUs. Does it use the cpu number ayway or the buffer number? @Andreagit97, do you know? :)

-------------------------------
*hits bong*

Now, hear me out. libscap is the absolutely worst place for suppressed tid logic anyway. If you go look through the code, it's literally the only place at this level that cares about the event schema. The other places are:
* drivers (obviously)
* libsinsp
* some helpers for the test_input engine IIRC

Everything else in libscap just shoves bytes around.

In my neverending refactoring work, I moved suppressed tids support to libsinsp (so we can at least use `std::unordered_set` instead of uthash) but maybe we want to move it to the engines instead (probably with a single shared implementation, at least initially).

This is not going to be beautiful, exactly, but then we can:
* add the cache to `scap_device`, where it can indeed be a single entry
* have a clean way forward to push down suppressions into the drivers (so the events never even get generated)
* support any number of devices (~cpus) since it's the driver's responsibility to allocate as many as it likes

The downside is that it will still be C so we can't have nice things like a decent set implementation :(
Another one would be that then we're tied to pcpuid as the cache slot selector ~forever.

Technically, we could come up with some monstrosity where higher layers can tack on data to devices, but it's probably overkill at this point. Or maybe:
* store just the (single entry) cache in devices
* provide a callback for "is this tid suppressed" from wherever (scap right now, sinsp eventually) for the engines to use and maintain the cache outside the engine (let's hope the function pointer caches well)
* eventually provide an engine vtable method like `notify_suppressed_tid(uint64_t tid, bool suppressed)` so that the engine can tell the driver about it and avoid generating events in the first place, but still the upper layer (scap -> sinsp) controls the logic. We'd have the least overhead if the driver implemented the same policy (suppress children of suppressed threads, unsuppress on exit) but it has the potential of bugs when the two implementations (driver/userspace) diverge

-------------------------------
*hits bong again*

With a memory allocation and an extra pointer field for each entry, it seems uthash is just about the heaviest data structure we could imagine for a set of numbers. Maybe some sparse bitmap would work better here? Something like https://github.com/RoaringBitmap/CRoaring?

Reply:
answered here https://github.com/falcosecurity/libs/pull/1159#pullrequestreview-1497133851

Reply:
> In my neverending refactoring work, I moved suppressed tids support to libsinsp (so we can at least use std::unordered_set instead of uthash) but maybe we want to move it to the engines instead

Yes, I total agree about moving it into the engines / drivers. Why? It moves the CPU overhead out of the agent main thread, and distributes it to all other threads. According to benchmarks, this could free up ~ 4.5% CPU in the agent main thread without this PR, and ~2% CPU in the agent main thread with this PR.

Problem is, whereas this PR is just a few lines of code, the future PR for moving the functionality into the engines / drivers would be way more complicated and more like open heart surgury... so that's why I view this PR as a "quick and dirty" in comparison, and a pragmatic stepping stone which is ultimately temporary, until the more complicated version is realized.

Reply:
> so we can at least use std::unordered_set instead of uthash

FYI when I started to look into optimizing the suppressed TIDs, then I optimized `uthash.h` to make it faster, and closer to the speed of e.g. `std::unordered_set` but... it still wasn't close to the performance of this PR. Why not? Because all hash table implementation -- including ` std::unordered_set` -- have two things in common: (a) there's always multple cache line fetches for many of the keys, and (b) there's always multiple key comparisions in a hash table bucket. Otherwise, the hash table would be a perfect hash table which can find the key and value immediately without searching through other keys in the bucket, etc.

But then I thought, why not try a different approach which is similar to the perfect hash table but without the "perfect" part. Since we know the events often come in "batches" from the same TID, why not take advantage of that knowledge and use the TID or CPUID (now DEVID in latest push to this PR) to go straight to the pre-hash table cache slot, and do a single compare to see if we have it cached or not. I also added some debug instrumentation to check how often there would be a cache hit or miss. And it came up with the following ratio:

```
debug: 1,280,000,000 = 1,279,716,383=scap_tid_cache_hit + 283,617=scap_tid_cache_miss
```

So we could avoid the hash table 1.279 billion times, using only the cache, and had to use the hash table only 283,617 times, or 0.022% of all lookups.

In this case, we get the fastest possible lookup of a fictional perfect hash table -- with a single cache line fetch, and simple comparison -- and we don't really care how quick or slow the underlying hash table implemenation is. We know the cache is always going to be the fastest for this application, and in theory we can save the work to swap out `uthash.h` for a different implementation, or?